### PR TITLE
p2p/enr: parse ENRs with enode

### PR DESCRIPTION
**Pull request by dB2510**
Corrects ENR encode/decode base64 format with the help of enode.

category: bug
ticket: #970

### CODE REVIEW

File: p2p/enr_test.go [hunk 1]
```diff
@@ -51,7 +51,13 @@ func TestDecodeENR_InvalidRLP(t *testing.T) {
...
 
 func TestDecodeENR_Oversize(t *testing.T) {
-	_, err := p2p.DecodeENR("enr:-IS4QBnEa-Oftjk7-sGRAY7IrvL5YjATdcHbqR5l2aXX2M25CiawfwaXh0k9hm98dCfdnqhz9mE-BfemFdjuL9KtHqgBgmlkgnY0gmlwhB72zxGJc2VjcDI1NmsxoQMaK8SspTrUgB8IYVI3qDgFYsHymPVsWlvIW477kxaKUIN0Y3CCJpUAAAA=")
+	_, err := p2p.DecodeENR("enr:-IS4QBnEa-Oftjk7-sGRAY7IrvL5YjATdcHbqR5l2aXX2M25CiawfwaXh0k9hm98dCfdnqhz9mE-BfemFdjuL9KtHqgBgmlkgnY0gmlwhB72zxGJc2VjcDI1NmsxoQMaK8SspTrUgB8IYVI3qDgFYsHymPVsWlvIW477kxaKUIN0Y3CCJpUAAAA")
 	require.Error(t, err)
 	require.Contains(t, err.Error(), "input contains more than one value")
 }
+
+func TestBackwardsENR(t *testing.T) {
+	oldVersionENR := "enr:-HW4QHwvU8usykhaKQ_cxyAaAck_TlS0zr9nLNXDODQll87ifLECDsnczJz7vC_aVTKq9Q8aC3UIKWwls1i4TmEXLAuAgmlkgnY0iXNlY3AyNTZrMaEDhry2ryLc-Y39laN92yfixHghkWIjLeQ4orBvDQy20jA="
```
Comment (corverroos) [inline p2p/enr_test.go:60]: better do 1000 random ENRs

Comment (corverroos) [reply p2p/enr_test.go:60]: also see encoding/base64/base64_test.go:62

**Review by corverroos — APPROVED** [review]

File: p2p/enr.go [hunk 2]
```diff
@@ -16,47 +16,36 @@
...
 func DecodeENR(enrStr string) (enr.Record, error) {
-	if enrStr == "" {
-		return enr.Record{}, errors.New("enr empty")
-	}
-	ok := strings.HasPrefix(enrStr, "enr:")
-	if !ok {
-		return enr.Record{}, errors.New("invalid ENR with no prefix (enr:)")
+	// Ensure backwards compatibility with older versions with padded ENR strings.
+	if strings.HasSuffix(enrStr, "=") {
+		enrStr = enrStr[:len(enrStr)-1]
```
Comment (corverroos) [inline p2p/enr.go:42]: this isn't correct

Comment (dB2510) [reply p2p/enr.go:42]: why?

Comment (dB2510) [reply p2p/enr.go:42]: got it

**Comment by corverroos** [comment]
Just adding "do not merge" until we fully understand why things were not breaking sooner

File: cmd/createcluster.go [hunk 3]
```diff
@@ -389,10 +390,29 @@ func newPeer(conf clusterConfig, peerIdx int) (p2p.Peer, error) {
 		return p2p.Peer{}, errors.Wrap(err, "create charon-enr-private-key")
 	}
 
-	var r enr.Record
-	err = enode.SignV4(&r, p2pKey)
+	p2pConfig := p2p.Config{
+		TCPAddrs: []string{"127.0.0.1:3610"},
+		UDPAddr:  "127.0.0.1:3630",
+	}
```
Comment (corverroos) [inline cmd/createcluster.go:396]: don't understand why this is required

File: p2p/enr.go [hunk 4]
```diff
@@ -16,47 +16,34 @@
...
 // See reference github.com/ethereum/go-ethereum@v1.10.10/p2p/dnsdisc/tree.go:378.
 func DecodeENR(enrStr string) (enr.Record, error) {
-	if enrStr == "" {
-		return enr.Record{}, errors.New("enr empty")
-	}
-	ok := strings.HasPrefix(enrStr, "enr:")
-	if !ok {
-		return enr.Record{}, errors.New("invalid ENR with no prefix (enr:)")
-	}
+	// Ensure backwards compatibility with older versions with encoded ENR strings.
```
Comment (corverroos) [inline p2p/enr.go:40]: add comment what that older actually did

File: p2p/enr_test.go [hunk 5]
```diff
@@ -51,7 +57,27 @@ func TestDecodeENR_InvalidRLP(t *testing.T) {
...
+	for i := 0; i < 1000; i++ {
+		_, record := testutil.RandomENR(t, random)
+		enrStr := encodeOldVersion(t, record)
+
+		_, err := p2p.DecodeENR(enrStr)
+		require.NoError(t, err)
+	}
+}
+
+func encodeOldVersion(t *testing.T, r enr.Record) string {
```
Comment (corverroos) [inline p2p/enr_test.go:76]: add comment what this is

File: p2p/enr_test.go [hunk 6]
```diff
@@ -51,7 +57,27 @@ func TestDecodeENR_InvalidRLP(t *testing.T) {
...
 func TestDecodeENR_Oversize(t *testing.T) {
-	_, err := p2p.DecodeENR("enr:-IS4QBnEa-Oftjk7-sGRAY7IrvL5YjATdcHbqR5l2aXX2M25CiawfwaXh0k9hm98dCfdnqhz9mE-BfemFdjuL9KtHqgBgmlkgnY0gmlwhB72zxGJc2VjcDI1NmsxoQMaK8SspTrUgB8IYVI3qDgFYsHymPVsWlvIW477kxaKUIN0Y3CCJpUAAAA=")
+	_, err := p2p.DecodeENR("enr:-IS4QBnEa-Oftjk7-sGRAY7IrvL5YjATdcHbqR5l2aXX2M25CiawfwaXh0k9hm98dCfdnqhz9mE-BfemFdjuL9KtHqgBgmlkgnY0gmlwhB72zxGJc2VjcDI1NmsxoQMaK8SspTrUgB8IYVI3qDgFYsHymPVsWlvIW477kxaKUIN0Y3CCJpUAAAA")
 	require.Error(t, err)
 	require.Contains(t, err.Error(), "input contains more than one value")
 }
+
+func TestBackwardsENR(t *testing.T) {
+	random := rand.New(rand.NewSource(time.Now().Unix()))
+	for i := 0; i < 1000; i++ {
```
Comment (corverroos) [inline p2p/enr_test.go:67]: 100 is probably fine

File: testutil/compose/define.go [hunk 7]
```diff
@@ -315,11 +316,28 @@ func copyStaticFolders(dir string) error {
 }
 
 func keyToENR(key *ecdsa.PrivateKey) (string, error) {
-	var r enr.Record
-	r.SetSeq(0)
+	config := p2p.Config{
+		TCPAddrs: []string{"127.0.0.1:3610"},
+		UDPAddr:  "127.0.0.1:3630",
+	}
```
Comment (corverroos) [inline testutil/compose/define.go:322]: doesn't seem right

**Review by corverroos — CHANGES_REQUESTED** [review]

**Review by corverroos — APPROVED** [review]

**Pull request merged**